src/main/java/com/uipath/uipathpackage/UiPathDeploy.java

```diff
@@ -106,37 +110,49 @@ public BuildStepMonitor getRequiredMonitorService() {
     /**
      * Run this step.
      *
-     * @param run      a build this is running as a part of
-     * @param filePath a workspace to use for any file operations
-     * @param launcher a way to start processes
-     * @param listener a place to send output
+     * @param run       a build this is running as a part of
+     * @param workspace a workspace to use for any file operations
+     * @param launcher  a way to start processes
+     * @param listener  a place to send output
      * @throws InterruptedException if the step is interrupted
      * @throws IOException          if something goes wrong
      */
     @Override
-    public void perform(@Nonnull Run<?, ?> run, @Nonnull FilePath filePath, @Nonnull Launcher launcher, @Nonnull TaskListener listener) throws InterruptedException, IOException {
-        EnvVars env = run.getEnvironment(listener);
-        ResourceBundle rb = ResourceBundle.getBundle("config");
+    public void perform(@Nonnull Run<?, ?> run, @Nonnull FilePath workspace, @Nonnull Launcher launcher, @Nonnull TaskListener listener) throws InterruptedException, IOException {
+        EnvVars envVars = run.getEnvironment(listener);
         Utility util = new Utility();
+        FilePath tempRemoteDir = tempDir(workspace);
+        tempRemoteDir.mkdirs();
+        ResourceBundle rb = ResourceBundle.getBundle("config");
+
         util.validateParams(orchestratorAddress, "Invalid Orchestrator Address");
         util.validateParams(packagePath, "Invalid Package Path");
-        String packagePathFormatted = escapePowerShellString(env.expand(packagePath.trim()));
-        String orchestratorTenantFormatted = env.expand(orchestratorTenant.trim()).isEmpty() ? util.getValue(rb, "UiPath.DefaultTenant") : escapePowerShellString(env.expand(orchestratorTenant.trim()));
+
+        FilePath packagePathFormatted = new FilePath(new File(envVars.expand(packagePath.trim())));
+        if (packagePathFormatted.isDirectory()) {
+            packagePathFormatted.copyRecursiveTo(tempRemoteDir);
+        } else {
+            packagePathFormatted.getParent().copyRecursiveTo(packagePathFormatted.getName(), tempRemoteDir); // can use copyTo instead of copyRecursiveTo
+        }
+
+        String orchestratorTenantFormatted = envVars.expand(orchestratorTenant.trim()).isEmpty() ? util.getValue(rb, "UiPath.DefaultTenant") : escapePowerShellString(envVars.expand(orchestratorTenant.trim()));
```

Comment:
envVars.expand is basically used for the 3 user inputs,
package path, orchestrator address, tenant name ( if provided, not required in case of default)
 escape powershell string is used in all inputs with 3 above inputs and additional for username and password.
So not in every place we need both escapepowershell and envVars.expand. It depends on the crux and nature of argument where we need each one.